add ad3530r driver

kseerp opened the pull request:
## PR Description

Part is unreleased, allowed by requestor to create PR for public release and initial review.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [X] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [X] I have conducted a self-review of my own code changes
- [X] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

kseerp commented:
v2
- address relevant CI complaints

nunojsa commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+	u8 ldac;
+	bool range_multiplier;
+
+	__be32 tx_buf __aligned(ARCH_DMA_MINALIGN);
```
this should be [IIO_DMA_MINALIGN](https://elixir.bootlin.com/linux/v6.12.6/C/ident/IIO_DMA_MINALIGN)

nunojsa commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+					 char *buf)
+{
+	struct ad3530r_state *st = iio_priv(indio_dev);
+
```
for correctness, you should also guard() your access in here

nunojsa commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+		return ret;
+
+	guard(mutex)(&st->lock);
+	st->chan[chan->channel].powerdown = powerdown;
```
This should only be set after a successful cal to ad3530r_reg_update_bits()

kseerp commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+#include <linux/module.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mutex.h>
+#include <linux/of.h>
```
this should not be included

kseerp commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+	.reg_bits = 16,
+	.val_bits = 8,
+	.read_flag_mask = BIT(7),
+};
```
I think I'm convinced not to use it due to multibyte registers, perhaps it would be better to use custom reg write and read

kseerp replied:
ohh my bad, regmap_bulk_read()/write() would definitely work with multibyte register, late to notice it. thanks for the info

kseerp commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+	IIO_ENUM("muxout_select", IIO_SHARED_BY_ALL, &ad3530r_muxout_select_enum),
+	IIO_ENUM_AVAILABLE("muxout_select", IIO_SHARED_BY_ALL,
+			   &ad3530r_muxout_select_enum),
+	{ },
```
I've seen existing drivers having muxout_select attr. I could also agree to make it a dt parameter based on your other comments. What do you suggest?

nunojsa replied:
Hmm, look at my other comments. As of now I think we should change this dynamically depending on what do we want to measure. I'm just not sure about the AGND mux...

kseerp commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+	[AD3530R_MUXOUT_IOUT7_SOURCE] = "iout7_source",
+	[AD3530R_MUXOUT_IOUT7_SINK] = "iout7_sink",
+	[AD3530R_MUXOUT_DIE_TEMP] = "die_temp",
+	[AD3530R_MUXOUT_AGND] = "agnd",
```
when in this config, MUX_OUT pin internally tied to AGND

nunojsa replied:
And what would be the usage for it? Just a configuration to no input is connected? So I guess we could set the muxout to this state after we done with some reading

kseerp commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,609 @@
+	[AD3530R_MUXOUT_VOUT6] = "vout6",
+	[AD3530R_MUXOUT_IOUT6_SOURCE] = "iout6_source",
+	[AD3530R_MUXOUT_IOUT6_SINK] = "iout6_sink",
+	[AD3530R_MUXOUT_VOUT7] = "vout7",
```
I think I might misunderstood it but all the channels are already exposed. The muxout_select attr is only used to select which of the mux input signals will be monitored on the mux_out pin.

nunojsa replied:
Nope, I'm the one who misunderstood things. I failed. to see you just have a template for the channels and then dynamically allocate all of the channels. From the above. the mux also does not work as I first guessed... From what you're saying it even seems to be something useful mainly for debugging? If so, you can also consider supporting it though debugFS. Anyways, up to you on how you want to present this attr upstream.

kseerp replied:
Yes, useful for debugging. I'll push another version soon for another round of review. Thanks

kseerp commented:
v3

- added ABI file

bindings:
- added iovcc-supply property
- made vcc-supply and iovcc-supply as required property
- added datasheet link

ad3530r:
- dropped unused macros, headers and registers
- used `devm_regulator_get_enable_read_voltage()`
- used IIO_DMA_MINALIGN macro for alignment
- used macros for hardcoded values
- modified state for reset-gpio

kseerp commented:
v4

* added support for the AD3531R (unreleased)
* introduced chip_info struct to handle number of channels and some different register addresses, such as input_ch_reg and sw_ldac_trig_reg
*  include new DAC powerdown mode (7.7kohm_to_gnd and 32kohm_to_gnd) in IIO ABI

will appreciate any reviews and feedback

kseerp commented:
v5

- utilize regmap API (bulk_read/writes for multibyte registers)

nunojsa commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,564 @@
+
+	return regmap_update_bits(st->regmap, st->chip_info->sw_ldac_trig_reg,
+				  AD3530R_SW_LDAC_TRIG_MASK,
+				  FIELD_PREP(AD3530R_SW_LDAC_TRIG_MASK, 1));
```
missing a lock

nunojsa commented on drivers/iio/dac/ad3530r.c:
```diff
@@ -0,0 +1,564 @@
+	for (i = 0; i < chip_info->num_channels; i++) {
+		st->channels[i] = ad3530r_channel_template;
+		st->channels[i].channel = i;
+	}
```
I wonder about the template usage... I'm not seeing anything being changed dynamically so it seems to be you could have a static channel definition.

kseerp replied:
Ohh I have in mind another device in ad353xr series that will be released soon. It has specific feature that can be included as extended info and can be changed dynamically through that implementation above

nunojsa commented:
Hi @kseerp , I see this was applied. Can you update this pr?

kseerp commented:
Hi @nunojsa, PR now updated. Thanks
